### PR TITLE
fix: BabyBear range check Gnark

### PR DESCRIPTION
**Pull request by erabinov**

### CODE REVIEW

File: recursion/circuit/src/fri.rs [hunk 1]
```diff
@@ -125,10 +129,17 @@ pub fn verify_two_adic_pcs<C: Config>(
...
+                            // Fill in any missing powers of alpha.
+                            (alpha_pows.len()..pow + 1).for_each(|_| {
+                                let new_alpha = builder.eval(*alpha_pows.last().unwrap() * alpha);
+                                builder.reduce_e(new_alpha);
+                                alpha_pows.push(new_alpha);
+                            });
+                            acc = builder.eval(acc + (alpha_pows[pow] * (p_at_z - p_at_x[0])));
+                            log_height_pow[log_height] += 1;
                         }
+                        // builder.reduce_e(acc);
```
Comment (erabinov) [inline recursion/circuit/src/fri.rs:142]:
```suggestion
```

File: recursion/circuit/src/stark.rs [hunk 2]
```diff
@@ -86,6 +86,8 @@ where
 
         let zeta = challenger.sample_ext(builder);
 
+        // builder.reduce_e(zeta);
+
```
Comment (erabinov) [inline recursion/circuit/src/stark.rs:90]:
```suggestion
```

File: recursion/circuit/src/stark.rs [hunk 3]
```diff
@@ -140,6 +142,7 @@ where
 
             let mut trace_points = Vec::new();
             let zeta_next = domain.next_point(builder, zeta);
+            // builder.reduce_e(zeta_next);
```
Comment (erabinov) [inline recursion/circuit/src/stark.rs:145]:
```suggestion
```

File: recursion/compiler/src/ir/utils.rs [hunk 4]
```diff
@@ -101,6 +101,22 @@ impl<C: Config> Builder<C> {
...
+    ) -> Ext<C::F, C::EF> {
+        let mut result = self.eval(SymbolicExt::from_f(C::EF::one()));
+        for i in 0..power_bits.len() {
+            let bit = power_bits[i];
+            let tmp = self.eval(result * two_adic_powers[i]);
+            result = self.select_ef(bit, tmp, result);
+        }
+        result
+    }
+
```
Comment (erabinov) [inline recursion/compiler/src/ir/utils.rs:119]:
```suggestion
```

File: recursion/circuit/src/fri.rs [hunk 5]
```diff
@@ -125,9 +129,15 @@ pub fn verify_two_adic_pcs<C: Config>(
                         let mut acc: Ext<C::F, C::EF> =
                             builder.eval(SymbolicExt::from_f(C::EF::zero()));
                         for (p_at_x, &p_at_z) in izip!(mat_opening.clone(), ps_at_z) {
-                            acc =
```
Comment (jtguibas) [inline recursion/circuit/src/fri.rs:128]: This performs the same logic. It just caches the power of alpha.

File: recursion/compiler/src/constraints/mod.rs [hunk 6]
```diff
@@ -358,6 +358,10 @@ impl<C: Config + Debug> ConstraintCompiler<C> {
                         vec![a[3].id()],
                     ],
                 }),
+                DslIr::ReduceE(a) => constraints.push(Constraint {
```
Comment (jtguibas) [inline recursion/compiler/src/constraints/mod.rs:361]: We added a reduction opcode to "force" a reduction.

**Review by jtguibas — APPROVED** [review]

**Pull request merged**